### PR TITLE
specify garbage collection

### LINKED ISSUE #13065: Should doc garbage collection strategy better
The garbage collection strategy has not been well doc (correct me if i'm wrong), which could bother users who have OOD issues badly.

My main concerns about gc comes to
- What would kubelet do with the volumes (hostPath, for instance) of the container precisely when it's gonna to remove a exited container. l'll bet It wont please the users whether kubelet cleans the volumes or not. Greater customization should be allowed here. 
- Should warn the users that an external garbage collection tool is not recommended since it would break kubelet's behavior.
- Since flag like `--maximum-dead-containers-per-container` has been provided to the users, what if we set it to 0? Would kubelet just accept this happily even it could hurt? 
- Is there still race condition (like break the pulling-image-process when gc) existing?

/cc @yujuhong  Seems you had a plan for this, i'm not so sure and just.. ding dong

### PR DESCRIPTION
**Pull request by dalanlan**
Fix  #13065

### CODE REVIEW

**Comment by dalanlan** [comment]
Copied from #13065 

My main concerns about gc comes to
- What would kubelet do with the volumes (hostPath, for instance) of the container precisely when it's gonna to remove a exited container. **l'll bet It wont please the users whether kubelet cleans the volumes or not. Greater customization should be allowed here. **
- Should warn the users that an external garbage collection tool is not recommended since it would break kubelet's behavior.
- **Since flag like `--maximum-dead-containers-per-container` has been provided to the users, what if we set it to 0? Would kubelet just accept this happily even it could hurt?**
- **Is there still race condition (like break the pulling-image-process when gc) existing?**
It would be nice to consider these for anyone who is going to review this pr.

**Comment by pwittrock** [comment]
I'll take a look.  Thanks Emma.

File: docs/user-guide/garbage-collection.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+Garbage collection is managed by kubelet automatically, mainly including unreferenced
+images and dead containers. kubelet applies container garbage collection every minute
+and image garbage collection every 5 minutes.
+
+### Image Collection
+
+kubernetes manages lifecycle of all images through `imageManager`。
+The policy for garbage collecting images we apply takes two factors into consideration,
+`HighThresholdPercent` and `LowThresholdPercent`. Disk usage above the the high threshold
+will trigger garbage collection,  and never down to low threshold.
```
Comment (pwittrock) [inline docs/user-guide/garbage-collection.md]: nit:
> ,  and never down to low threshold.
Is this strictly true?  Garbage collection attempts to free space by deleting images until "usage - (int64(im.policy.LowThresholdPercent) * capacity / 100)" bytes are freed.  Maybe something like "Disk usage above the the high threshold will trigger garbage collection, which will attempt to delete unused images until the `LowThresholdPercent` is met"?

nit:
Maybe also mention that least recently used images are deleted first.

File: docs/user-guide/garbage-collection.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+> Note that we will skip the containers that are not managed by kubelet.
+
+### User Configuration
+
+Users are free to set their own value to address image garbage collection.
+
+- `image-gc-high-threshold`, the percent of disk usage after which image garbage collection is
+always run. Default is 90%
+- `image-gc-low-threshold`, the percent of disk usage before which image garbage collection is
```
Comment (pwittrock) [inline docs/user-guide/garbage-collection.md]: > the percent of disk usage before which image garbage collection is never run

This reads a bit awkward to me.  Is this used for anything other than determining how much space to attempt to free up?  If not, maybe just say something like "garbage collection will attempt to free space until this percent of disk is free"

File: docs/user-guide/garbage-collection.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+never run. Lowest disk usage to garbage collect to. Default is 80%.
+
+We also allow users to customize garbage collection policy, basically via following three flags.
+
+- `minimum-container-ttl-duration`, minimum age for a finished container before it is garbage collected.
+- `maximum-dead-containers-per-container`, maximum number of old instances of a container to retain
+per container. Each container takes up some disk space.
+- `maximum-dead-containers`, maximum number of old instances of a containers to retain globally. Each
+container takes up some disk space.
+Value less than zero for the last two is regarded as no limit.
```
Comment (pwittrock) [inline docs/user-guide/garbage-collection.md]: super nit:  I believe these should be plural e.g.
Values less than zero for the last two are regarded as no limit.

File: docs/user-guide/garbage-collection.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+kubernetes manages lifecycle of all images through `imageManager`。
+The policy for garbage collecting images we apply takes two factors into consideration,
+`HighThresholdPercent` and `LowThresholdPercent`. Disk usage above the the high threshold
+will trigger garbage collection,  and never down to low threshold.
+
+### Container Collection
+
+The policy for garbage collecting containers we apply takes on three variables, which can
+be user-defined. See also [here](#User Configuration).
+kubelet sorts out containers which are unidentified or stay out of bounds set by previous
```
Comment (pwittrock) [inline docs/user-guide/garbage-collection.md]: Possibly mention the order in which the flags are enforced: e.g.
1. maximum-dead-containers-per-container
2. maximum-dead-containers

**Comment by pwittrock** [comment]
Hey Emma,

with respect to the concerns you listed, are you concerned about what the current behavior is and how to document it, or are you asking how we can improve the existing behavior?

File: docs/user-guide/garbage-collection.md (hunk 2 above)
Comment (dalanlan) [inline docs/user-guide/garbage-collection.md]: Hmm.. These are copied from usage of the params directly (from the code). Consider it taken care of:)

File: docs/user-guide/garbage-collection.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Garbage Collection
```
Comment (bgrant0607) [inline docs/user-guide/garbage-collection.md]: Please move this doc to docs/admin and add a link to the README.md in that directory, as a sub-bullet under "The kubelet binary"

**Comment by bgrant0607** [comment]
cc @kubernetes/goog-node

**Comment by bgrant0607** [comment]
cc @vishh re. disk management

File: docs/user-guide/garbage-collection.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+1. `minimum-container-ttl-duration`, minimum age for a finished container before it is 
+garbage collected.
+2. `maximum-dead-containers-per-container`, maximum number of old instances of a container to retain
+per container.
+3. `maximum-dead-containers`, maximum number of old instances of a containers to retain globally. Each
+container takes up some disk space.
+Values less than zero for the last two are regarded as no limit.
+
+> Note that we don't recommend external garbage collection tool, since it could break the behavior
```
Comment (bgrant0607) [inline docs/user-guide/garbage-collection.md]: Was the leading `>` intentional?

Comment (dalanlan) [inline docs/user-guide/garbage-collection.md]: It was. Deprecated now (though).

**Comment by dalanlan** [comment]
> with respect to the concerns you listed, are you concerned about what the current behavior is and how to document it, or are you asking how we can improve the existing behavior?

Mostly the latter one.

> What would kubelet do with the volumes (hostPath, for instance) of the container precisely when it's gonna to remove a exited container. l'll bet It wont please the users whether kubelet cleans the volumes or not. Greater customization should be allowed here.

Cleaning up all of the volumes roughly seems a little unfriendly to me. What do you say?
> Since flag like `--maximum-dead-containers-per-container` has been provided to the users, what if we set it to 0? Would kubelet just accept this happily even it could hurt?

kubelet relies on the dead containers to make the strategy sort of. Hence zero should be a unwanted param value theoretically. 
 
> Is there still race condition (like break the pulling-image-process when gc) existing?

This is simply a question. Inspired by #3393

/cc @pwittrock  The comments has been addressed. Please have a double check:)

**Comment by dalanlan** [comment]
Fix #8416

**Comment by pwittrock** [comment]
@dalanlan Thanks for writing this documentation up.  I am sure folks will find it very helpful.

> What would kubelet do with the volumes (hostPath, for instance) of the container precisely when it's gonna to remove a exited container. l'll bet It wont please the users whether kubelet cleans the volumes or not. Greater customization should be allowed here.

I am still learning about how GC interacts with volumes, but @vishh may be able to give a better answer here.  If this is something that you would like to discuss more, it probably makes sense to open a separate issue to do so.

> Should warn the users that an external garbage collection tool is not recommended since it would break kubelet's behavior.

I agree we should warn users about this.  I saw that you did so in your doc.  Thanks!

>Since flag like --maximum-dead-containers-per-container has been provided to the users, what if we set it to 0? Would kubelet just accept this happily even it could hurt? kubelet relies on the dead containers to make the strategy sort of. Hence zero should be a unwanted param value theoretically, yet permitted now.

Good point about relying on dead containers, this would be a bug.  From looking at the code, is does look like the GC will accept a flag of 0 for dead-per-container, and perform the garbage collection.  We should probably create an issue to address bugs caused by garbage collecting containers we rely upon for correctly obeying restart policy.  I am not certain that setting the min flag value to 1 will solve this issue entirely because the flag for max total dead containers may cause the same issue.  I would consider flag values that cause undesirable but correct behavior (such as poor performance or deleting containers the use may want to examine later) to be a separate and more complicated issue.

> Is there still race condition (like break the pulling-image-process when gc) existing?

I am not aware of any known race conditions caused by container or image garbage collection, and didn't see any open issues after doing a simple search.  @yujuhong can you confirm?

File: docs/admin/garbage-collection.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Garbage Collection
+
+Garbage collection is managed by kubelet automatically, mainly including unreferenced
+images and dead containers. kubelet applies container garbage collection every minute
+and image garbage collection every 5 minutes.
+Note that we don't recommend external garbage collection tool, since it could break
```
Comment (vishh) [inline docs/admin/garbage-collection.md]: A side note: Docker still can leak files and forget about them. External garbage collection tools to clean up behind docker might be useful. WDYT?

File: docs/admin/garbage-collection.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+The policy for garbage collecting containers we apply takes on three variables, which can
+be user-defined. `MinAge` is the minimum age at which a container can be garbage collected,
+zero for no limit. `MaxPerPodContainer` is the max number of dead containers any single
+pod (UID, container name) pair is allowed to have, less than zero for no limit.
+`MaxContainers` is the max number of total dead containers, less than zero for no limit.
+
+kubelet sorts out containers which are unidentified or stay out of bounds set by previous
+mentioned three flags. For the latter case, the oldest containers are removed first.
+When kubelet cleans up the dead containers, the associated volumes will be taken care of as well.
```
Comment (vishh) [inline docs/admin/garbage-collection.md]: Are you referring to the files written by the container or to pod-scoped volumes?

File: docs/admin/garbage-collection.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+Users are free to set their own value to address image garbage collection.
+
+1. `image-gc-high-threshold`, the percent of disk usage which triggers image garbage collection.
+Default is 90%.
+2. `image-gc-low-threshold`, the percent of disk usage to which image garbage collection attempts
+to free. Default is 80%.
+
+We also allow users to customize garbage collection policy, basically via following three flags.
+
+1. `minimum-container-ttl-duration`, minimum age for a finished container before it is
```
Comment (vishh) [inline docs/admin/garbage-collection.md]: What are the defaults for these flags?

File: docs/admin/garbage-collection.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+garbage collected.
+2. `maximum-dead-containers-per-container`, maximum number of old instances of a container to retain
+per container.
+3. `maximum-dead-containers`, maximum number of old instances of a containers to retain globally. Each
```
Comment (vishh) [inline docs/admin/garbage-collection.md]: nit: remove 's' from `containers`

File: docs/admin/garbage-collection.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+1. `image-gc-high-threshold`, the percent of disk usage which triggers image garbage collection.
+Default is 90%.
+2. `image-gc-low-threshold`, the percent of disk usage to which image garbage collection attempts
+to free. Default is 80%.
+
+We also allow users to customize garbage collection policy, basically via following three flags.
+
+1. `minimum-container-ttl-duration`, minimum age for a finished container before it is
+garbage collected.
+2. `maximum-dead-containers-per-container`, maximum number of old instances of a container to retain
```
Comment (vishh) [inline docs/admin/garbage-collection.md]: nit: remove `of a container`.

**Comment by vishh** [comment]
1. We generally discourage using hostpath for storing data. Use a volume. What if your container were to run on a machine with no local disk? May be, we need to document the recommendation clearly. 
The more the knobs the harder it is to understand and maintain the system.
2. +1 for warning. As I mentioned in a comment, docker might still leak, and we might need an external garbage collector, or make kubelet smart enough to handle that.
3. Not storing previous dead containers will affect debugging - mainly access to logs. On a node with very little disk space, it is possible for users to not retain any old containers, except for the running ones. WDYT? Our APIs should fail gracefully, if previous instances of a container is not found.
4. I would hope for docker to handle these races. A image pull and deletion that ends up touching the same layers, should be synchronized by docker.

File: docs/admin/garbage-collection.md (hunk 7 above)
Comment (dalanlan) [inline docs/admin/garbage-collection.md]: The leaking issues troubles us a lot as well, i'll admit it, yet it's (maybe) a long-term run pb.

File: docs/admin/garbage-collection.md (hunk 8 above)
Comment (dalanlan) [inline docs/admin/garbage-collection.md]: Exactly.

**Comment by dalanlan** [comment]
The comments have been addressed.  @vishh 
>We generally discourage using hostpath for storing data. Use a volume. 

Yet hostPath is a type of volume though. More knobs can cost a lot indeed OTOH. Better doc would be an easy way.
>Our APIs should fail gracefully, if previous instances of a container is not found.
The concern comes to retaining dead container issue was inspired by [this comment](https://github.com/kubernetes/kubernetes/issues/12856#issuecomment-132703278), we have good reasons to believe that kubelet is not quite itself w/o intended tombstone.
It would be improved by other pr like checkpoint(#489), yet it still could hurt with a improper flag in the short term.

File: docs/admin/garbage-collection.md (hunk 8 above)
Comment (vishh) [inline docs/admin/garbage-collection.md]: Ok. Can you rephrase `volumes` as the `write layer of container`?

**Comment by vishh** [comment]
I wonder if we can get rid of the container and somehow just preserve the log files in docker. Another approach is to provide access to historical logs only via a cluster logging framework. Sadly, none of these mechanisms exist today.

**Comment by vishh** [comment]
@dalanlan: Thanks for the pointer to #12856. I get your comment now. We do have to enforce a policy around the gc flags.

File: docs/admin/garbage-collection.md (hunk 8 above)
Comment (dalanlan) [inline docs/admin/garbage-collection.md]: Hmm..sorry for my carelessness. A container-scoped one.

Comment (dalanlan) [inline docs/admin/garbage-collection.md]: `write layer of container` sounds a bit confusing to me, to be honest. 
Is this term like the layer beyond an image thing when running a container?

**Comment by pwittrock** [comment]
I've filed #13287 to track
- Retaining container state after a container has been garbage collected
- Ensuring access to log data after a container has died

File: docs/admin/garbage-collection.md (hunk 8 above)
Comment (vishh) [inline docs/admin/garbage-collection.md]: How about `all the files inside the container will be cleaned up as well`? By write-layer I was referring to the write layer that docker adds to all containers, which is mapped to some directory on the host filesystem.

Comment (dalanlan) [inline docs/admin/garbage-collection.md]: Done :)
I was wondering whether the rephrasing job is aimed to get rid of ambiguity of `volume`.

**Comment by dalanlan** [comment]
The comments have been fully addressed.
/cc @pwittrock  @vishh

File: docs/admin/garbage-collection.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+1. `image-gc-high-threshold`, the percent of disk usage which triggers image garbage collection.
+Default is 90%.
+2. `image-gc-low-threshold`, the percent of disk usage to which image garbage collection attempts
+to free. Default is 80%.
+
+We also allow users to customize garbage collection policy, basically via following three flags.
+
+1. `minimum-container-ttl-duration`, minimum age for a finished container before it is
+garbage collected. Default is 1 minute.
+2. `maximum-dead-containers-per-container`, maximum number of old instances to retain
```
Comment (vishh) [inline docs/admin/garbage-collection.md]: Should we add a line stating that this should be set to a minimum value of `2` as of now?

**Comment by vishh** [comment]
Just one more comment @dalanlan! Otherwise LGTM. 
@pwittrock I will let you make a final approval!

File: docs/admin/garbage-collection.md (hunk 12 above)
Comment (dalanlan) [inline docs/admin/garbage-collection.md]: It's kinda tricky. What do you say about `maximum-dead-containers` flag then? That one could hurt as well :-)

Comment (vishh) [inline docs/admin/garbage-collection.md]: Yes. Thats true. I was thinking that providing hints to have a stable
system will be very helpful to users. WDYT?

Comment (pwittrock) [inline docs/admin/garbage-collection.md]: It is probably worth mentioning that maximum-dead-containers should be large enough to allow at least 2 dead containers-per-expected-container?  May be also reference #13287 as an explanation for why these values are recommended.

**Comment by dalanlan** [comment]
I'd updated the doc.  
1. Add the logic of policy dealing with the case in which `maximum-dead-containers-per-container` and `maximum-dead-containers` conflict. 
2. Sort out the previous mentioned flags pb.
PTAL /cc @vishh  @pwittrock

**Comment by pwittrock** [comment]
@dalanlan Thanks for all your work on this

**Comment by dalanlan** [comment]
Shame. I cannot really access the test details for Jenkins. And it should not break anything to change doc.
May trigger this again, please? @pwittrock

**Comment by pwittrock** [comment]
@k8s-bot test this again please
